synapse/storage/schema/main/delta/78/01_validate_and_update_profiles.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2023 The Matrix.org Foundation C.I.C
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from synapse.config.homeserver import HomeServerConfig
+from synapse.storage.database import LoggingTransaction
+from synapse.storage.engines import BaseDatabaseEngine, PostgresEngine
+
+
+def run_upgrade(
+    cur: LoggingTransaction,
+    database_engine: BaseDatabaseEngine,
+    config: HomeServerConfig,
+) -> None:
+    hostname = config.server.server_name
+
+    if isinstance(database_engine, PostgresEngine):
+        # check if the constraint can be validated
+        check_sql = """
+                    SELECT user_id from profiles WHERE full_user_id IS NULL
+                    """
+        cur.execute(check_sql)
+        res = cur.fetchall()
+
+        if res:
+            # there are rows the background job missed, finish them here before we validate the constraint
+            process_rows_sql = """
+                               UPDATE profiles
+                               SET full_user_id = '@' || user_id || ?
+                               WHERE user_id IN (SELECT user_id FROM profiles WHERE
+                               full_user_id IS NULL)
```

Comment:
Do we have an appropriate index to make this `WHERE full_user_id IS NULL` performant?

Reply:
There are indexes on both `full_user_id` and `user_id` (as there is a `UNIQUE` constraint on both of them) so I think it should be fine?

Reply:
:+1: I think the unique index is sufficient. [Postgres](https://www.postgresql.org/docs/current/indexes-unique.html) and [SQLite](https://www.sqlite.org/lang_createindex.html) consider `NULL`'s to be unique so they will all be part of the index. And specifically, it looks like Postgres will store whether something is [`NULL` in a bitmap](https://dba.stackexchange.com/a/210294/138493) before the values in the index. And from the query analysis, it seems like it's using exactly that bitmap (`Bitmap Index Scan`):

```
$ psql --user postgres synapse
> EXPLAIN ANALYZE SELECT *
FROM profiles
WHERE full_user_id is NULL;
                                                            QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on profiles  (cost=4.17..11.27 rows=3 width=128) (actual time=0.005..0.005 rows=0 loops=1)
   Recheck Cond: (full_user_id IS NULL)
   ->  Bitmap Index Scan on profiles_full_user_id_key  (cost=0.00..4.17 rows=3 width=0) (actual time=0.003..0.003 rows=0 loops=1)
         Index Cond: (full_user_id IS NULL)
 Planning Time: 0.060 ms
 Execution Time: 0.019 ms
(6 rows)
```

Reply:
Awesome thank for the in-depth explanation!